File: code/jumpgame1.java
class Solution {
    public boolean canJump(int[] nums) {
        int max_jump = 0;
        int n = nums.length;

        for(int i=0; i<n; i++){
            if(nums[i] == 0 && i < n-1 && max_jump <= i){
                return false;
            }else if(max_jump < (i+nums[i])){
                max_jump = i+nums[i];
            }
        }
        
        return max_jump >= n-1;
    }
}

File: code/jumpgame1.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        int dp[n];

        for(int i=0; i<n; i++){
            dp[i] = 0;
        }

        dp[n-1] = 1;

        for(int i=n-2; i>=0; i--){
            int min_id = min(n-1, nums[i]+i);
            for(int j=i; j<=min_id; j++){
                if(dp[j] == 1){
                    dp[i] = 1;
                    break;
                }
            }
            

        }
        
        return dp[0] == 1;
    }
};
